sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala

```diff
@@ -173,4 +174,29 @@ private case object OracleDialect extends JdbcDialect {
     val nullable = if (isNullable) "NULL" else "NOT NULL"
     s"ALTER TABLE $tableName MODIFY ${quoteIdentifier(columnName)} $nullable"
   }
+
+  override def getLimitClause(limit: Integer): String = {
+    // Oracle doesn't support LIMIT clause.
+    // We can use rownum <= n to limit the number of rows in the result set.
+    if (limit > 0 ) s"WHERE rownum <= $limit" else ""
+  }
+
+  class OracleSQLQueryBuilder(dialect: JdbcDialect, options: JDBCOptions)
+    extends JdbcSQLQueryBuilder(dialect, options) {
+
+    // TODO[SPARK-42289]: DS V2 pushdown could let JDBC dialect decide to push down offset
+    override def build(): String = {
+      val selectStmt = s"SELECT $columnList FROM ${options.tableOrQuery} $tableSampleClause" +
+        s" $whereClause $groupByClause $orderByClause"
+      if (limit > 0) {
```

Comment:
Thanks for fixing the limit. Could you add a test for Oracle JDBC limit? Those integration tests don't test limit, otherwise we would have encountered this issue before.

Reply:
The two test case `simple scan with LIMIT` and `simple scan with top N` are in the parent trait.